qiskit/visualization/utils.py

```diff
@@ -208,5 +138,166 @@ def _get_gate_span(qregs, instruction):
 
     if instruction.cargs:
         return qregs[min_index:]
+    if instruction.condition:
+        return qregs[min_index:]
 
     return qregs[min_index:max_index + 1]
+
+
+def _any_crossover(qregs, node, nodes):
+    """Return True .IFF. 'node' crosses over any in 'nodes'
+    qiskit-terra #2802
+    """
+    gate_span = _get_gate_span(qregs, node)
+    all_indices = []
+    for check_node in nodes:
+        if check_node != node:
+            all_indices += _get_gate_span(qregs, check_node)
+    return any(i in gate_span for i in all_indices)
+
+
+class LayerSpooler():
+    """Manipulate list of layer dicts for _get_layered_instructions
+    qiskit-terra #2802
+    """
+
+    def __init__(self, dag, justification):
+        """Create spool"""
+        self.dag = dag
+        self.qregs = dag.qubits()
+        self.justification = justification
+        self.spool = []
+
+        if self.justification == 'left':
+
+            for dag_layer in dag.layers():
+                current_index = self.last_index()
+                dag_nodes = _sorted_nodes(dag_layer)
+                for node in dag_nodes:
+                    self.add(node, current_index)
+
+        else:
+            dag_layers = []
+
+            for dag_layer in dag.layers():
+                dag_layers.append(dag_layer)
+
+            # going right to left!
+            dag_layers.reverse()
+
+            for dag_layer in dag_layers:
+                current_index = 0
+                dag_nodes = _sorted_nodes(dag_layer)
+                for node in dag_nodes:
+                    self.add(node, current_index)
+
+    def size(self):
+        """Return number of entries in spool"""
+        return len(self.spool)
+
+    def last_index(self):
+        """Return last index in spool"""
+        return len(self.spool) - 1
+
+    def as_list(self):
+        """Get the spool as a list of layer values"""
+        return list(self.spool)
+
+    def append(self, layer):
+        """Append to spool"""
+        self.spool.append(layer)
```

Comment:
`LayerSpooler` is not a design contribution to the conceptual architecture of Qiskit : it is merely a consistent, readable, and maintainable piece of code which replaces two mutually incompatible and incorrect spaghetti functions :)

`LayerSpooler` is very clear about what it does and what each individual method of the class does. As it has no relevance outside the context of `utils.py`, I would suggest the easiest way forward is to make all its member functions private.

Since it works correctly, exacts no excessive performance penalty, and is tidy in itself, though admittedly no masterpiece of Pythonic elegance, it seems to me any coding time I would spend in further design, rework and testing of this already working and already tested piece of code could be more usefully expended on another issue.

I'll be glad to prepend underscores to the names of all member functions if that will simplify matters ;)

Reply:
> I'll be glad to prepend underscores to the names of all member functions if that will simplify matters ;)

Let me know if this is acceptable to resolve your concerns and I will then do it.

Reply:
Made _LayerSpooler private.

Reply:
I pushed up https://github.com/Qiskit/qiskit-terra/pull/3061/commits/040da0b4636f90032a8f37caad09e85cff7cd511 which makes _LayerSpooler a subclass of list which addresses most of my concerns about this.